File: strings/0003-longestSubstring.py
"""
leetcode 3 longest substring
double-pointer: left and right
"""

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic, res, i = {}, 0, -1
        for j in range(len(s)):
            if s[j] in dic:
                i = max(dic[s[j]], i) # update left pointer i
            dic[s[j]] = j # record in the dictionary as hash-table
            res = max(res, j - i) # update the result
        return res

File: strings/0003-longestSubstring.cpp
/*
 * leetcode 3
 * longest substring
 *
 * use unordered_set as the hash-table
 * 
 */

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        int len = s.size();
        int window = 0;
        unordered_set<char> uset;
        for (int l = 0, r = 0; r < len; r++) {
            while (l < r && uset.count(s[r])) {
                // s[r] already in uset
                // so [l, r] has duplicates
                // now move the left pointet so that there is no duplicates
                // in the window [l, r]
                uset.erase(s[l]);
                l++;
                window--;
            }
            uset.insert(s[r]);
            window++;
            res = max(res, window);
        }    
        return res;
    }
};

